server/src/main/java/org/candlepin/policy/SystemPurposeComplianceRules.java

```diff
@@ -0,0 +1,214 @@
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * ComplianceRules
+ *
+ * A class used to check consumer compliance status.
+ */
+public class SystemPurposeComplianceRules {
+    private static Logger log = LoggerFactory.getLogger(SystemPurposeComplianceRules.class);
+
+    private EntitlementCurator entCurator;
+    private StatusReasonMessageGenerator generator;
+    private EventSink eventSink;
+    private ConsumerCurator consumerCurator;
+    private ConsumerTypeCurator consumerTypeCurator;
+    private I18n i18n;
+
+    @Inject
+    public SystemPurposeComplianceRules(EntitlementCurator entCurator,
+        StatusReasonMessageGenerator generator, EventSink eventSink, ConsumerCurator consumerCurator,
+        ConsumerTypeCurator consumerTypeCurator, I18n i18n) {
+
+        this.entCurator = entCurator;
+        this.generator = generator;
+        this.eventSink = eventSink;
+        this.consumerCurator = consumerCurator;
+        this.consumerTypeCurator = consumerTypeCurator;
+        this.i18n = i18n;
+    }
+
+    /**
+     * Check compliance status for a consumer on a specific date.
+     *
+     * @param consumer Consumer to check.
+     * @param updateConsumer whether or not to use consumerCurator.update
+     *        (also expensive)
+     * @return Compliance status.
+     */
+    @SuppressWarnings("checkstyle:indentation")
+    public SystemPurposeComplianceStatus getStatus(Consumer consumer,
+        Collection<Entitlement> existingEntitlements, Collection<Entitlement> newEntitlements,
+        boolean updateConsumer, boolean currentCompliance) {
+
+        SystemPurposeComplianceStatus status = new SystemPurposeComplianceStatus(i18n);
+
+        if (existingEntitlements == null) {
+            existingEntitlements = new LinkedList<>();
+        }
+
+        if (newEntitlements == null) {
+            newEntitlements = new LinkedList<>();
+        }
+
+        Set<Entitlement> entitlements = Stream.concat(
+            existingEntitlements.stream(),
+            newEntitlements.stream())
+            .collect(Collectors.toSet());
+
+        for (Entitlement entitlement : entitlements) {
+
+            String unsatisfedRole = consumer.getRole();
+            Set<String> unsatisfiedAddons = new HashSet<>(consumer.getAddOns());
+            String preferredSla = consumer.getServiceLevel();
+            String preferredUsage = consumer.getUsage();
+
+            Set<Product> products = Stream.concat(
+                entitlement.getPool().getProvidedProducts().stream(),
+                entitlement.getPool().getDerivedProvidedProducts().stream()
+            ).collect(Collectors.toSet());
+            products.add(entitlement.getPool().getProduct());
+            Product derivedProduct = entitlement.getPool().getDerivedProduct();
```

Comment:
This worries me a bit because it's not deterministic.  I can't tell for sure what implementation `Collectors.toSet()` is using, but I'd like it if we used a `TreeSet` (items are sorted by comparator) or `LinkedHashSet` (items are returned in insertion order).

Reply:
Sorting can be expensive. why is the order important?

Reply:
It's probably not a big deal.  I was wondering at the time whether it would mean that the status detail message the consumer gets would just be randomly ordered.  Re-reading, I don't think this is a concern, so please disregard.